google/cloud/spanner_dbapi/transaction_helper.py

```diff
@@ -0,0 +1,105 @@
+# limitations under the License.
+
+from typing import TYPE_CHECKING, List
+
+import time
+from google.api_core.exceptions import Aborted
+
+from google.cloud.spanner_dbapi.utils import PeekIterator
+
+if TYPE_CHECKING:
+    from google.cloud.spanner_dbapi import Connection
+from google.cloud.spanner_dbapi.checksum import ResultsChecksum, _compare_checksums
+from google.cloud.spanner_dbapi.parsed_statement import Statement
+from google.cloud.spanner_v1.session import _get_retry_delay
+
+from google.rpc.code_pb2 import ABORTED
+
+MAX_INTERNAL_RETRIES = 50
+
+
+class TransactionHelper:
+    def __init__(self, connection: "Connection"):
+        self._connection = connection
+        # Non-Batch statements, which were executed within the current
+        # transaction
+        self._single_statements: List[Statement] = []
+        # Batch statements, which were executed within the current transaction
+        self._batch_statements_list: List[(List[Statement], ResultsChecksum)] = []
+
+    def retry_transaction(self):
+        """Retry the aborted transaction.
+
+        All the statements executed in the original transaction
+        will be re-executed in new one. Results checksums of the
+        original statements and the retried ones will be compared.
+
+        :raises: :class:`google.cloud.spanner_dbapi.exceptions.RetryAborted`
+            If results checksum of the retried statement is
+            not equal to the checksum of the original one.
+        """
+        attempt = 0
+        while True:
+            self._connection._spanner_transaction_started = False
+            attempt += 1
+            if attempt > MAX_INTERNAL_RETRIES:
+                raise
+
+            try:
+                self._rerun_previous_statements()
+                self._rerun_previous_batch_statements()
+                break
+            except Aborted as exc:
+                delay = _get_retry_delay(exc.errors[0], attempt)
+                if delay:
+                    time.sleep(delay)
+
+    def _rerun_previous_batch_statements(self):
+        """
+        Helper to run all the remembered statements from the last transaction.
+        """
+        for batch_statements, original_checksum in self._batch_statements_list:
+            transaction = self._connection.transaction_checkout()
+            statements_tuple = []
+            for single_statement in batch_statements:
+                statements_tuple.append(single_statement.get_tuple())
+            status, res = transaction.batch_update(statements_tuple)
+            if status.code == ABORTED:
+                raise Aborted(status.details)
+
+            retried_checksum = _get_batch_statements_result_checksum(res, status.code)
+            _compare_checksums(original_checksum, retried_checksum)
+
+    def _rerun_previous_statements(self):
+        for single_statement in self._single_statements:
+            res_iter = self._connection.run_statement(single_statement)
```

Comment:
If this statement fails (for example if the statement returns `syntax error`), then the entire retry will fail with syntax error. That is not correct, and would also be very confusing to the user, as they would get a `syntax error` message for a random statement that triggered the retry.

It is important to consider that the list of statements in the transaction can contain statements that will never work, for example because they contain a syntax error (or any other error). If the user caught those exceptions during the initial transaction attempt, then they will (should!) be in the list of statements, and should also be retried. During the retry, it should be verified that the statement returned the same error as during the initial attempt.

In Java we for example have the class `FailedUpdate` that is recorded as the executed statement if a DML statement fails during the initial attempt. During the retry, the connection checks that the same error code is returned as during the initial attempt: https://github.com/googleapis/java-spanner/blob/c56c56981b00ee059ec4dbad2154f24ebc12f8b9/google-cloud-spanner/src/main/java/com/google/cloud/spanner/connection/FailedUpdate.java#L32

Reply:
> It is important to consider that the list of statements in the transaction can contain statements that will never work, for example because they contain a syntax error (or any other error). If the user caught those exceptions during the initial transaction attempt, then they will (should!) be in the list of statements, and should also be retried. During the retry, it should be verified that the statement returned the same error as during the initial attempt.

I am not adding the failed statement with exception being caught by user to be retried. Shouldn't that be ok as the user already caught the exception for those statements?

Reply:
Statements that return an error are part of the original attempt of the transaction, and must also be part of the retry attempt. The reason for this is that an error during a transaction can be an indication of what is and is not present in the database.

Consider for example a very naive implementation of insert-or-update (in pseudo-code):

```sql
try:
  insert into foo (id, value) values (1, 'One');
  return;
rescue AlreadyExistsError:
  update foo set value='One' where id=1;
```

During a retry we should verify that `insert into foo` returns the same result as during the initial attempt, as that result can be used to control the further program flow. A 'result' in this case can also be a specific error.